corehq/motech/fhir/models.py

```diff
@@ -0,0 +1,169 @@
+from jsonfield import JSONField
+
+from corehq.apps.data_dictionary.models import CaseType, CaseProperty
+from corehq.motech.exceptions import ConfigurationError
+from corehq.motech.fhir.const import FHIR_VERSIONS, FHIR_VERSION_4_0_1
+from corehq.motech.value_source import ValueSource, as_value_source
+
+
+@attr.s(auto_attribs=True)
+class PropertyInfo:
+    name: str
+    json_name: str
+    type_: type
+    is_list: bool
+    of_many: Optional[str]
+    is_required: bool
+
+
+class FHIRResourceType(models.Model):
+    domain = models.CharField(max_length=127, db_index=True)
+    fhir_version = models.CharField(max_length=12, choices=FHIR_VERSIONS,
+                                    default=FHIR_VERSION_4_0_1)
+    case_type = models.ForeignKey(CaseType, on_delete=models.CASCADE)
+
+    # `fhirclient_class` values look like `module_name.ClassName`, and
+    # can be imported from fhirclient.models.*module_name.ClassName*
+    fhirclient_class = models.CharField(max_length=255)
+
+    # `template` is used for defining a JSON document structure if it
+    # cannot be built using only FHIRResourceAttributes
+    template = JSONField(null=True, blank=True, default=None)
+
+    def get_fhirclient_class(self) -> type:
+        """
+        Returns a FHIR resource class from the fhirclient library.
+
+        >>> resource_type = FHIRResourceType(
+        ...     case_type=CaseType(name='contact'),
+        ...     fhirclient_class='patient.PatientContact',
+        ... )
+        >>> class_type = resource_type.get_fhirclient_class()
+        >>> class_type.__name__
+        'PatientContact'
+
+        """
+        try:
+            module_name, class_name = self.fhirclient_class.split('.')
+            module = import_module(f'fhirclient.models.{module_name}')
+            return getattr(module, class_name)
+        except (ValueError, ImportError, AttributeError) as err:
+            raise ConfigurationError('Unknown FHIR resource type '
+                                     f'{self.fhirclient_class!r}') from err
+
+    def get_properties_info(self) -> List[PropertyInfo]:
+        """
+        Returns a list of info about each resource property.
+
+        >>> resource_type = FHIRResourceType(
+        ...     case_type=CaseType(name='contact'),
+        ...     fhirclient_class='patient.Patient',
+        ... )
+        >>> for pi in resource_type.get_properties_info():
+        ...    if pi.of_many == 'multipleBirth':
+        ...        print(pi.json_name, pi.type_)
+        multipleBirthBoolean <class 'bool'>
+        multipleBirthInteger <class 'int'>
+
+        (For more information about how FHIR represents multiple births,
+        see FHIR `Patient`_ documentation.)
+
+        .. _Patient: https://www.hl7.org/fhir/patient.html
+
+        """
+        class_type = self.get_fhirclient_class()
+        return [PropertyInfo(*p) for p in class_type().elementProperties()]
+
+
+class FHIRResourceProperty(models.Model):
```

Comment:
This seems like it is going to cause many joins for every resource. Do you think that will be a problem? I guess we can optimize these out with `select_related` when making queries?

Reply:
Yes, I definitely think we need to use `select_related`.